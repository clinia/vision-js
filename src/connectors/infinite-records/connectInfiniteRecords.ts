import {
  CliniaSearchHelper as Helper,
  SearchParameters,
} from '@clinia/search-helper';
import {
  Renderer,
  RendererOptions,
  WidgetFactory,
  Records,
  Unmounter,
} from '../../types';
import {
  checkRendering,
  createDocumentationMessageGenerator,
  isEqual,
  addAbsolutePosition,
  addQueryID,
  noop,
} from '../../lib/utils';
import { InfiniteRecordsRendererWidgetParams } from '../../widgets/infinite-records/infinite-records';

export type InfiniteRecordsConnectorParams = Partial<
  InfiniteRecordsRendererWidgetParams
>;

export interface InfiniteRecordsRendererOptions<TInfiniteRecordsWidgetParams>
  extends RendererOptions<TInfiniteRecordsWidgetParams> {
  showPrevious: () => void;
  showMore: () => void;
  isFirstPage: boolean;
  isLastPage: boolean;
}

export type InfiniteRecordsRenderer<TInfiniteRecordsWidgetParams> = Renderer<
  InfiniteRecordsRendererOptions<
    InfiniteRecordsConnectorParams & TInfiniteRecordsWidgetParams
  >
>;

export type InfiniteRecordsWidgetFactory<
  TInfiniteRecordsWidgetParams
> = WidgetFactory<
  InfiniteRecordsConnectorParams & TInfiniteRecordsWidgetParams
>;

export type InfiniteRecordsConnector = <TInfiniteRecordsWidgetParams>(
  render: InfiniteRecordsRenderer<TInfiniteRecordsWidgetParams>,
  unmount?: Unmounter
) => InfiniteRecordsWidgetFactory<TInfiniteRecordsWidgetParams>;

const withUsage = createDocumentationMessageGenerator({
  name: 'infinite-records',
  connector: true,
});

const connectInfiniteRecords: InfiniteRecordsConnector = (
  renderFn,
  unmountFn = noop
) => {
  checkRendering(renderFn, withUsage());

  return widgetParams => {
    const {
      transformItems = (items: any[]) => items,
      showPrevious: hasShowPrevious = false,
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    } = widgetParams || ({} as typeof widgetParams);
    let recordsCache: Records = [];
    let firstReceivedPage = Infinity;
    let lastReceivedPage = -1;
    let prevState: Partial<SearchParameters>;
    let showPrevious: () => void;
    let showMore: () => void;

    const getShowPrevious = (helper: Helper): (() => void) => () => {
      // Using the helper's `overrideStateWithoutTriggeringChangeEvent` method
      // avoid updating the browser URL when the user displays the previous page.
      helper
        .overrideStateWithoutTriggeringChangeEvent({
          ...helper.state,
          page: firstReceivedPage - 1,
        })
        .search();
    };
    const getShowMore = (helper: Helper): (() => void) => () => {
      helper.setPage(lastReceivedPage + 1).search();
    };
    const filterEmptyRefinements = (refinements = {}) => {
      return Object.keys(refinements)
        .filter(key =>
          Array.isArray(refinements[key])
            ? refinements[key].length
            : Object.keys(refinements[key]).length
        )
        .reduce((obj, key) => {
          obj[key] = refinements[key];
          return obj;
        }, {});
    };

    return {
      $$type: 'cvi.infiniteRecords',

      init({ visionInstance, helper }) {
        showPrevious = getShowPrevious(helper);
        showMore = getShowMore(helper);
        firstReceivedPage = helper.state.page || 0;
        lastReceivedPage = helper.state.page || 0;

        renderFn(
          {
            records: recordsCache,
            results: undefined,
            showPrevious,
            showMore,
            isFirstPage: firstReceivedPage === 0,
            isLastPage: true,
            visionInstance,
            widgetParams,
          },
          true
        );
      },

      render({ results, state, visionInstance }) {
        // Reset cache and received pages if anything changes in the
        // search state, except for the page.
        //
        // We're doing this to "reset" the widget if a refinement or the
        // query changes between renders, but we want to keep it as is
        // if we only change pages.
        const {
          page = 0,
          facets,
          disjunctiveFacets,
          maxValuesPerFacet,
          ...currentState
        } = state;

        currentState.facetsRefinements = filterEmptyRefinements(
          currentState.facetsRefinements
        );
        currentState.disjunctiveFacetsRefinements = filterEmptyRefinements(
          currentState.disjunctiveFacetsRefinements
        );

        if (!isEqual(currentState, prevState)) {
          recordsCache = [];
          firstReceivedPage = page;
          lastReceivedPage = page;
          prevState = currentState;
        }

        results.records = addAbsolutePosition(
          results.records,
          results.page,
          results.perPage
        );

        results.records = addQueryID(results.records, results.queryID);

        results.records = transformItems(results.records);

        if (lastReceivedPage < page || !recordsCache.length) {
          recordsCache = [...recordsCache, ...results.records];
          lastReceivedPage = page;
        } else if (firstReceivedPage > page) {
          recordsCache = [...results.records, ...recordsCache];
          firstReceivedPage = page;
        }

        const isFirstPage = firstReceivedPage === 0;
        const isLastPage = results.numPages <= results.page + 1;

        renderFn(
          {
            records: recordsCache,
            results,
            showPrevious,
            showMore,
            isFirstPage,
            isLastPage,
            visionInstance,
            widgetParams,
          },
          false
        );
      },

      dispose({ state }) {
        unmountFn();

        const stateWithoutPage = state.setQueryParameter('page', undefined);

        return stateWithoutPage;
      },

      getWidgetState(uiState, { searchParameters }) {
        const page = searchParameters.page || 0;

        if (!hasShowPrevious || !page) {
          return uiState;
        }

        return {
          ...uiState,
          // The page in the UI state is incremented by one
          // to expose the user value (not `0`).
          page: page + 1,
        };
      },

      getWidgetSearchParameters(searchParameters, { uiState }) {
        const widgetSearchParameters = searchParameters;

        // The page in the search parameters is decremented by one
        // to get to the actual parameter value from the UI state.
        const page = uiState.page ? uiState.page - 1 : 0;

        return widgetSearchParameters.setQueryParameter('page', page);
      },
    };
  };
};

export default connectInfiniteRecords;
