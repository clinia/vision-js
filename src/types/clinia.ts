import { SearchClient as SearchClientV2 } from 'clinia';
import { SearchResponse as SearchResponseV2 } from '@clinia/client-search';

export type Client = SearchClientV2;

export interface MultiResponse<TRecord = any> {
  results: Array<SearchResponse<TRecord>>;
}

export type SearchResponse<TRecord> = SearchResponseV2<TRecord>;
