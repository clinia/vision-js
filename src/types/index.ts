export * from './clinia';
export * from './connector';
export * from './insights';
export * from './vision';
export * from './widget';
