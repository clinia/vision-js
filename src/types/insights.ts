export type InsightsClientMethod =
  | 'clickedRecordIDsAfterSearch'
  | 'convertedRecordIDsAfterSearch';

export type InsightsClientPayload = {
  eventName: string;
  queryID: string;
  index: string;
  ids: string[];
  positions?: number[];
};

export type InsightsClient = (
  method: InsightsClientMethod,
  payload: InsightsClientPayload
) => void;

export type InsightsClientWrapper = (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => void;
