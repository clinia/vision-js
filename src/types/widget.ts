import { Index } from '../widgets/index/index';
import {
  CliniaSearchHelper as Helper,
  SearchParameters,
  SearchResults,
  PlainSearchParameters,
} from '@clinia/search-helper';
import { Vision } from './vision';

export interface InitOptions {
  visionInstance: Vision;
  parent: Index | null;
  uiState: UiState;
  state: SearchParameters;
  helper: Helper;
  templatesConfig: object;
  createURL(state: SearchParameters): string;
}

export interface ScopedResult {
  indexId: string;
  results: SearchResults;
  helper: Helper;
}

export interface RenderOptions {
  visionInstance: Vision;
  templatesConfig: object;
  results: SearchResults;
  scopedResults: ScopedResult[];
  state: SearchParameters;
  helper: Helper;
  searchMetadata: {
    isSearchStalled: boolean;
  };
  createURL(state: SearchParameters): string;
}

export interface DisposeOptions {
  helper: Helper;
  state: SearchParameters;
}

export interface WidgetStateOptions {
  searchParameters: SearchParameters;
  helper: Helper;
}

export interface WidgetSearchParametersOptions {
  uiState: IndexUiState;
}

export type IndexUiState = {
  query?: string;
  refinementList?: {
    [property: string]: string[];
  };
  menu?: {
    [property: string]: string;
  };

  toggle?: {
    [property: string]: boolean;
  };
  geoSearch?: {
    /**
     * The rectangular area in geo coordinates.
     * The rectangle is defined by two diagonally opposite points, hence by 4 floats separated by commas.
     *
     * @example '47.3165,4.9665,47.3424,5.0201'
     */
    boundingBox: string;
  };
  page?: number;
  perPage?: number;
  configure?: PlainSearchParameters;
  places?: {
    query: string;
    /**
     * The central geolocation.
     *
     * @example '48.8546,2.3477'
     */
    position: string;
  };
};

export type UiState = {
  [indexId: string]: IndexUiState;
};

/**
 * Widgets are the building blocks of Visionjs. Any valid widget must
 * have at least a `render` or a `init` function.
 */
export interface Widget {
  $$type?:
    | 'cvi.autocomplete'
    | 'cvi.breadcrumb'
    | 'cvi.clearRefinements'
    | 'cvi.configure'
    | 'cvi.configureRelatedItems'
    | 'cvi.currentRefinements'
    | 'cvi.geoSearch'
    | 'cvi.hierarchicalMenu'
    | 'cvi.records'
    | 'cvi.perPage'
    | 'cvi.index'
    | 'cvi.infiniteRecords'
    | 'cvi.menu'
    | 'cvi.numericMenu'
    | 'cvi.pagination'
    | 'cvi.places'
    | 'cvi.poweredBy'
    | 'cvi.queryRules'
    | 'cvi.queryRuleCustomData'
    | 'cvi.queryRuleContext'
    | 'cvi.range'
    | 'cvi.rangeInput'
    | 'cvi.rangeSlider'
    | 'cvi.ratingMenu'
    | 'cvi.refinementList'
    | 'cvi.searchBox'
    | 'cvi.sortBy'
    | 'cvi.stats'
    | 'cvi.toggleRefinement'
    | 'cvi.voiceSearch';
  /**
   * Called once before the first search
   */
  init?(options: InitOptions): void;
  /**
   * Called after each search response has been received
   */
  render?(options: RenderOptions): void;
  /**
   * Called when this widget is unmounted. Used to remove refinements set by
   * during this widget's initialization and life time.
   */
  dispose?(options: DisposeOptions): SearchParameters | void;
  /**
   * This function is required for a widget to be taken in account for routing.
   * It will derive a uiState for this widget based on the existing uiState and
   * the search parameters applied.
   * @param uiState current state
   * @param widgetStateOptions extra information to calculate uiState
   */
  getWidgetState?(
    uiState: IndexUiState,
    widgetStateOptions: WidgetStateOptions
  ): IndexUiState;
  /**
   * This function is required for a widget to behave correctly when a URL is
   * loaded via e.g. routing. It receives the current UiState and applied search
   * parameters, and is expected to return a new search parameters.
   * @param state applied search parameters
   * @param widgetSearchParametersOptions extra information to calculate next searchParameters
   */
  getWidgetSearchParameters?(
    state: SearchParameters,
    widgetSearchParametersOptions: WidgetSearchParametersOptions
  ): SearchParameters;
}

export type WidgetFactory<TWidgetParams> = (
  widgetParams: TWidgetParams
) => Widget;

export type Template<TTemplateData = void> =
  | string
  | ((data: TTemplateData) => string);
