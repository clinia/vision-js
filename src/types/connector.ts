import { SearchResults } from '@clinia/search-helper';
import { Records, Vision } from './vision';
import { InsightsClient } from './insights';

export type RendererOptions<TWidgetParams = unknown> = {
  /**
   * Original parameters for this widget.
   * Useful for giving back the render parameters to the renderer.
   */
  widgetParams: TWidgetParams;
  visionInstance: Vision;
  results?: SearchResults;
  records?: Records;
  insights?: InsightsClient;
};

export type Renderer<TRenderOptions extends RendererOptions = any> = (
  renderOptions: TRenderOptions,
  isFirstRender: boolean
) => void;

export type Unmounter = () => void;
