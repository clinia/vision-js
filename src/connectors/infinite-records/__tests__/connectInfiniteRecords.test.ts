import searchHelper, {
  SearchResults,
  SearchParameters,
} from '@clinia/search-helper';
import { Client } from '../../../types';
import { createVision } from '../../../../test/mock/createVision';
import {
  createInitOptions,
  createRenderOptions,
} from '../../../../test/mock/createWidget';
import { createSingleSearchResponse } from '../../../../test/mock/createAPIResponse';
import connectInfiniteRecords from '../connectInfiniteRecords';
import { createSearchClient } from '../../../../test/mock/createSearchClient';

jest.mock('../../../lib/utils/records-absolute-position', () => ({
  // The real implementation creates a new array instance, which can cause bugs,
  // especially with the __escaped mark, we thus make sure the mock also has the
  // same behavior regarding the array.
  addAbsolutePosition: records => records.map(x => x),
}));

describe('connectInfiniteRecords', () => {
  it('throws without render function', () => {
    expect(() => {
      // @ts-ignore: test connectInfiniteRecords with invalid parameters
      connectInfiniteRecords()({});
    }).toThrowErrorMatchingInlineSnapshot(`
"The render function is not valid (received type Undefined).

See documentation: https://www.clinia.com/doc/api-reference/widgets/infinite-records/js/#connector"
`);
  });

  it('is a widget', () => {
    const render = jest.fn();
    const unmount = jest.fn();

    const customInfiniteRecords = connectInfiniteRecords(render, unmount);
    const widget = customInfiniteRecords({});

    expect(widget).toEqual(
      expect.objectContaining({
        $$type: 'cvi.infiniteRecords',
        init: expect.any(Function),
        render: expect.any(Function),
        dispose: expect.any(Function),

        getWidgetState: expect.any(Function),
        getWidgetSearchParameters: expect.any(Function),
      })
    );
  });

  it('Renders during init and render', () => {
    const renderFn = jest.fn();
    const visionInstance = createVision();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({
      escapeHTML: true,
    });

    expect(renderFn).toHaveBeenCalledTimes(0);

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        visionInstance,
        state: helper.state,
        helper,
      })
    );

    expect(renderFn).toHaveBeenCalledTimes(1);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        visionInstance,
        records: [],
        showPrevious: expect.any(Function),
        showMore: expect.any(Function),
        results: undefined,
        isFirstPage: true,
        isLastPage: true,
        widgetParams: {
          escapeHTML: true,
        },
      }),
      true
    );

    widget.render!(
      createRenderOptions({
        results: new SearchResults(helper.state, [
          createSingleSearchResponse({ records: [] }),
        ]),
        state: helper.state,
        visionInstance,
        helper,
      })
    );

    expect(renderFn).toHaveBeenCalledTimes(2);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        visionInstance,
        records: [],
        showPrevious: expect.any(Function),
        showMore: expect.any(Function),
        results: expect.any(Object),
        isFirstPage: true,
        isLastPage: true,
        widgetParams: {
          escapeHTML: true,
        },
      }),
      false
    );
  });

  it('Provides the records and accumulates results on next page', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    const firstRenderOptions = renderFn.mock.calls[0][0];
    expect(firstRenderOptions.records).toEqual([]);
    expect(firstRenderOptions.results).toBe(undefined);

    const records = [
      { fake: 'data', id: '1' },
      { sample: 'infos', id: '2' },
    ];
    const results = new SearchResults(helper.state, [
      createSingleSearchResponse({ records }),
    ]);

    widget.render!(
      createRenderOptions({
        state: helper.state,
        results,
        helper,
      })
    );

    const secondRenderOptions = renderFn.mock.calls[1][0];
    const { showMore } = secondRenderOptions;
    expect(secondRenderOptions.records).toEqual(records);
    expect(secondRenderOptions.results).toEqual(results);
    showMore();
    expect(helper.search).toHaveBeenCalledTimes(1);

    // the results should accumulate if there is an increment in page
    const otherRecords = [
      { fake: 'data 2', id: '1' },
      { sample: 'infos 2', id: '2' },
    ];
    const otherResults = new SearchResults(helper.state, [
      createSingleSearchResponse({ records: otherRecords }),
    ]);

    widget.render!(
      createRenderOptions({
        results: otherResults,
        state: helper.state,
        helper,
      })
    );

    const thirdRenderOptions = renderFn.mock.calls[2][0];
    expect(thirdRenderOptions.records).toEqual([...records, ...otherRecords]);
    expect(thirdRenderOptions.results).toEqual(otherResults);
  });

  it('Provides the records and prepends results on previous page', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({} as Client, '', {});
    helper.setPage(1);
    helper.search = jest.fn();
    helper.emit = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    const firstRenderOptions = renderFn.mock.calls[0][0];
    expect(firstRenderOptions.records).toEqual([]);
    expect(firstRenderOptions.results).toBe(undefined);

    const records = [
      { fake: 'data', id: '1' },
      { sample: 'infos', id: '2' },
    ];
    const results = new SearchResults(helper.state, [
      createSingleSearchResponse({ records }),
    ]);

    widget.render!(
      createRenderOptions({
        state: helper.state,
        results,
        helper,
      })
    );

    const secondRenderOptions = renderFn.mock.calls[1][0];
    const { showPrevious } = secondRenderOptions;
    expect(secondRenderOptions.records).toEqual(records);
    expect(secondRenderOptions.results).toEqual(results);
    showPrevious();
    expect(helper.state.page).toBe(0);
    expect(helper.emit).not.toHaveBeenCalled();
    expect(helper.search).toHaveBeenCalledTimes(1);

    // the results should be prepended if there is an decrement in page
    const previousRecords = [
      { fake: 'data 2', id: '1' },
      { sample: 'infos 2', id: '2' },
    ];
    const previousResults = new SearchResults(helper.state, [
      createSingleSearchResponse({ records: previousRecords }),
    ]);

    widget.render!(
      createRenderOptions({
        results: previousResults,
        state: helper.state,
        helper,
      })
    );

    const thirdRenderOptions = renderFn.mock.calls[2][0];
    expect(thirdRenderOptions.records).toEqual([
      ...previousRecords,
      ...records,
    ]);
    expect(thirdRenderOptions.results).toEqual(previousResults);
  });

  it('Provides the records and flush hists cache on query changes', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    const firstRenderOptions = renderFn.mock.calls[0][0];
    expect(firstRenderOptions.records).toEqual([]);
    expect(firstRenderOptions.results).toBe(undefined);

    const records = [
      { fake: 'data', id: '1' },
      { sample: 'infos', id: '2' },
    ];
    const results = new SearchResults(helper.state, [
      createSingleSearchResponse({ records }),
    ]);

    widget.render!(
      createRenderOptions({
        state: helper.state,
        results,
        helper,
      })
    );

    const secondRenderOptions = renderFn.mock.calls[1][0];
    expect(secondRenderOptions.records).toEqual(records);
    expect(secondRenderOptions.results).toEqual(results);

    helper.setQuery('data');

    // If the query changes, the records cache should be flushed
    const otherRecords = [
      { fake: 'data 2', id: '1' },
      { sample: 'infos 2', id: '2' },
    ];
    const otherResults = new SearchResults(helper.state, [
      createSingleSearchResponse({ records: otherRecords }),
    ]);

    widget.render!(
      createRenderOptions({
        results: otherResults,
        state: helper.state,
        helper,
      })
    );

    const thirdRenderOptions = renderFn.mock.calls[2][0];
    expect(thirdRenderOptions.records).toEqual(otherRecords);
    expect(thirdRenderOptions.results).toEqual(otherResults);
  });

  it('transform items if requested', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({
      transformItems: items => items.map(() => ({ name: 'transformed' })),
    });

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    const firstRenderOptions = renderFn.mock.calls[0][0];
    expect(firstRenderOptions.records).toEqual([]);
    expect(firstRenderOptions.results).toBe(undefined);

    const records = [
      {
        name: 'name 1',
        id: '1',
      },
      {
        name: 'name 2',
        id: '2',
      },
    ];

    const results = new SearchResults(helper.state, [
      createSingleSearchResponse({ records }),
    ]);

    widget.render!(
      createRenderOptions({
        state: helper.state,
        results,
        helper,
      })
    );

    const transformedRecords = [
      {
        name: 'transformed',
      },
      {
        name: 'transformed',
      },
    ];

    const secondRenderOptions = renderFn.mock.calls[1][0];
    expect(secondRenderOptions.records).toEqual(transformedRecords);
    expect(secondRenderOptions.results).toEqual(results);
  });

  it('adds queryID if provided to results', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    const records = [
      {
        name: 'name 1',
        id: '1',
      },
      {
        name: 'name 2',
        id: '2',
      },
    ];

    const results = new SearchResults(helper.state, [
      createSingleSearchResponse({ records, queryID: 'theQueryID' }),
    ]);

    widget.render!(
      createRenderOptions({
        state: helper.state,
        results,
        helper,
      })
    );

    expect(renderFn).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        records: [
          {
            name: 'name 1',
            id: '1',
            __queryID: 'theQueryID',
          },
          {
            name: 'name 2',
            id: '2',
            __queryID: 'theQueryID',
          },
        ],
      }),
      false
    );
  });

  it('does not render the same page twice', () => {
    const renderFn = jest.fn();
    const makeWidget = connectInfiniteRecords(renderFn);
    const widget = makeWidget({});

    const helper = searchHelper({} as Client, '', {});
    helper.search = jest.fn();

    widget.init!(
      createInitOptions({
        state: helper.state,
        helper,
      })
    );

    widget.render!(
      createRenderOptions({
        results: new SearchResults(helper.state, [
          createSingleSearchResponse({
            records: [{ id: 'a' }],
            page: 0,
            numPages: 4,
          }),
        ]),
        state: helper.state,
        helper,
      })
    );

    helper.setPage(1);

    widget.render!(
      createRenderOptions({
        results: new SearchResults(helper.state, [
          createSingleSearchResponse({
            records: [{ id: 'b' }],
            page: 1,
            numPages: 4,
          }),
        ]),
        state: helper.state,
        helper,
      })
    );

    expect(renderFn).toHaveBeenCalledTimes(3);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        records: [{ id: 'a' }, { id: 'b' }],
      }),
      false
    );

    widget.render!(
      createRenderOptions({
        results: new SearchResults(helper.state, [
          createSingleSearchResponse({
            records: [{ id: 'b' }],
            page: 1,
            numPages: 4,
          }),
        ]),
        state: helper.state,
        helper,
      })
    );

    expect(renderFn).toHaveBeenCalledTimes(4);
    expect(renderFn).toHaveBeenLastCalledWith(
      expect.objectContaining({
        records: [{ id: 'a' }, { id: 'b' }],
      }),
      false
    );
  });

  describe('dispose', () => {
    it('calls the unmount function', () => {
      const helper = searchHelper({} as Client, '', {});

      const renderFn = (): void => {};
      const unmountFn = jest.fn();
      const makeWidget = connectInfiniteRecords(renderFn, unmountFn);
      const widget = makeWidget({});

      expect(unmountFn).toHaveBeenCalledTimes(0);

      widget.dispose!({ helper, state: helper.state });

      expect(unmountFn).toHaveBeenCalledTimes(1);
    });

    it('does not throw without the unmount function', () => {
      const helper = searchHelper({} as Client, '', {});

      const renderFn = (): void => {};
      const makeWidget = connectInfiniteRecords(renderFn);
      const widget = makeWidget({});

      expect(() =>
        widget.dispose!({ helper, state: helper.state })
      ).not.toThrow();
    });

    it('does not remove the TAG_PLACEHOLDER from the `SearchParameters` with `escapeHTML` disabled', () => {
      const helper = searchHelper({} as Client, '', {
        highlightPreTag: '<mark>',
        highlightPostTag: '</mark>',
      });

      const renderFn = (): void => {};
      const makeWidget = connectInfiniteRecords(renderFn);
      const widget = makeWidget({
        escapeHTML: false,
      });

      expect(helper.state.highlightPreTag).toBe('<mark>');
      expect(helper.state.highlightPostTag).toBe('</mark>');

      const nextState = widget.dispose!({
        helper,
        state: helper.state,
      }) as SearchParameters;

      expect(nextState.highlightPreTag).toBe('<mark>');
      expect(nextState.highlightPostTag).toBe('</mark>');
    });

    it('removes the `page` from the `SearchParameters`', () => {
      const helper = searchHelper({} as Client, '', {
        page: 5,
      });

      const renderFn = (): void => {};
      const makeWidget = connectInfiniteRecords(renderFn);
      const widget = makeWidget({});

      expect(helper.state.page).toBe(5);

      const nextState = widget.dispose!({
        helper,
        state: helper.state,
      }) as SearchParameters;

      expect(nextState.page).toBeUndefined();
    });
  });

  describe('getWidgetState', () => {
    test('returns the `uiState` empty without `showPrevious` option', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName', {
        page: 1,
      });
      const widget = makeWidget({});

      const actual = widget.getWidgetState!(
        {},
        {
          searchParameters: helper.state,
          helper,
        }
      );

      expect(actual).toEqual({});
    });

    test('returns the `uiState` empty with `showPrevious` option on first page', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName', {
        page: 0,
      });
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetState!(
        {},
        {
          searchParameters: helper.state,
          helper,
        }
      );

      expect(actual).toEqual({});
    });

    test('returns the `uiState` containing `page` with `showPrevious` option', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName', {
        page: 1,
      });
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetState!(
        {},
        {
          searchParameters: helper.state,
          helper,
        }
      );

      expect(actual).toEqual({
        page: 2,
      });
    });

    test('returns the `uiState` containing a page number incremented by one with `showPrevious` option and `page` search parameter', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName', {
        page: 3,
      });
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetState!(
        {},
        {
          searchParameters: helper.state,
          helper,
        }
      );

      expect(actual).toEqual({
        page: 4,
      });
    });
  });

  describe('getWidgetSearchParameters', () => {
    test('return Search Parameters without highlighted tags when `escapeHTML` is `false`', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName');
      const widget = makeWidget({
        escapeHTML: false,
      });

      const actual = widget.getWidgetSearchParameters!(helper.state, {
        uiState: {},
      });

      expect(actual.highlightPreTag).toBeUndefined();
      expect(actual.highlightPostTag).toBeUndefined();
    });

    test('return Search Parameters with resetted page when `showPrevious` without `page` in the UI state', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName');
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetSearchParameters!(helper.state, {
        uiState: {},
      });

      expect(actual.page).toEqual(0);
    });

    test('return Search Parameters with resetted page when `showPrevious` and `page` is 0 in the UI state', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName');
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetSearchParameters!(helper.state, {
        uiState: {
          page: 0,
        },
      });

      expect(actual.page).toEqual(0);
    });

    test('return Search Parameters with page decreased when `showPrevious` and `page` in the UI state', () => {
      const render = jest.fn();
      const makeWidget = connectInfiniteRecords(render);
      const helper = searchHelper(createSearchClient(), 'indexName');
      const widget = makeWidget({
        showPrevious: true,
      });

      const actual = widget.getWidgetSearchParameters!(helper.state, {
        uiState: {
          page: 3,
        },
      });

      expect(actual.page).toEqual(2);
    });
  });
});
