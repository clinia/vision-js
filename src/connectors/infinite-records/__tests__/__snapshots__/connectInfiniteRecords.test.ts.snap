// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectInfiniteRecords throws without render function 1`] = `
"The render function is not valid (received type Undefined).

See documentation: https://www.clinia.com/doc/api-reference/widgets/infinite-records/js/#connector"
`;
